### PR TITLE
Make Pools not be Accounts

### LINKED ISSUE #113: Check to confirm that PoolAccount addresses are ADR 028 compliant
If they aren't, we should either make them ADR 028 compliant, or remove the addresses for now.

### PR DESCRIPTION
**Pull request by sunnya97**
Pools themselves shouldn't be accounts.  Instead, they have an Address field that corresponds to a specific ModuleAccount in the accountkeeper.

These addresses should be ADR-028 compliant.

closes #113

### CODE REVIEW

**Comment by sunnya97** [comment]
Needed to import some code from sdk v0.43 for the adr-028 compliance.  This can be removed when we upgrade to v0.43.

File: app/params/address.go [hunk 1]
```diff
@@ -35,4 +37,21 @@ func SetAddressPrefixes() {
 	config.SetBech32PrefixForAccount(Bech32PrefixAccAddr, Bech32PrefixAccPub)
 	config.SetBech32PrefixForValidator(Bech32PrefixValAddr, Bech32PrefixValPub)
 	config.SetBech32PrefixForConsensusNode(Bech32PrefixConsAddr, Bech32PrefixConsPub)
+
+	config.SetAddressVerifier(func(bytes []byte) error {
```
Comment (ValarDragon) [inline app/params/address.go:41]: Add a comment that this is copied from the SDK?

File: proto/osmosis/gamm/v1beta1/pool.proto [hunk 2]
```diff
@@ -109,15 +109,12 @@ message PoolParams {
...
   option (gogoproto.goproto_getters) = false;
   option (gogoproto.goproto_stringer) = false;
-  option (cosmos_proto.implements_interface) = "PoolAccountI";
+  option (cosmos_proto.implements_interface) = "PoolI";
 
-  cosmos.auth.v1beta1.BaseAccount baseAccount = 1 [
-    (gogoproto.embed) = true,
-    (gogoproto.moretags) = "yaml:\"base_account\""
-  ];
+  string address = 1 [ (gogoproto.moretags) = "yaml:\"future_pool_governor\"" ];
```
Comment (ValarDragon) [inline proto/osmosis/gamm/v1beta1/pool.proto:117]: Incorrect tag

File: x/gamm/types/pool.pb.go [hunk 3]
```diff
@@ -223,10 +223,10 @@ func (m *PoolParams) GetSmoothWeightChangeParams() *SmoothWeightChangeParams {
 	return nil
 }
 
-type PoolAccount struct {
-	*types1.BaseAccount `protobuf:"bytes,1,opt,name=baseAccount,proto3,embedded=baseAccount" json:"baseAccount,omitempty" yaml:"base_account"`
-	Id                  uint64     `protobuf:"varint,2,opt,name=id,proto3" json:"id,omitempty"`
-	PoolParams          PoolParams `protobuf:"bytes,3,opt,name=poolParams,proto3" json:"poolParams" yaml:"pool_params"`
+type Pool struct {
+	Address    string     `protobuf:"bytes,1,opt,name=address,proto3" json:"address,omitempty" yaml:"future_pool_governor"`
```
Comment (ValarDragon) [inline x/gamm/types/pool.pb.go:227]: Incorrect yaml

File: x/gamm/types/pool.go [hunk 4]
```diff
@@ -40,30 +42,28 @@ type PoolAccountI interface {
...
-	MaxUserSpecifiedWeight    sdk.Int      = sdk.NewIntFromUint64(1 << 20)
-	GuaranteedWeightPrecision int64        = 1 << 30
+	_                         PoolI   = (*Pool)(nil)
+	MaxUserSpecifiedWeight    sdk.Int = sdk.NewIntFromUint64(1 << 20)
+	GuaranteedWeightPrecision int64   = 1 << 30
 )
 
 func NewPoolAddress(poolId uint64) sdk.AccAddress {
-	return sdk.AccAddress(crypto.AddressHash(append(PoolAddressPrefix, sdk.Uint64ToBigEndian(poolId)...)))
+	return address.Module(ModuleName, sdk.Uint64ToBigEndian(poolId))
```
Comment (ValarDragon) [inline x/gamm/types/pool.go:51]: Lets add more domain separation here. I suggest making the RHS `"Pool" + sdk.Uint64ToBigEndian(poolId)`

Comment (ValarDragon) [reply x/gamm/types/pool.go:51]: (The concern is that there could one day be other addresses in this module, and then we have a headache for ensuring that they can never conflict)

File: x/gamm/types/pool.go [hunk 5]
```diff
@@ -144,27 +144,37 @@ func (params PoolParams) Validate(poolWeights []PoolAsset) error {
 	return nil
 }
 
-func (pa PoolAccount) GetId() uint64 {
+// GetAddress returns the address of a pool.
+// If the pool address is not bech32 valid, it returns an empty address.
```
Comment (ValarDragon) [inline x/gamm/types/pool.go:148]: Should we panic instead? Why would a pool not have a poolID (which then implies an address)

File: x/gamm/types/pool.go [hunk 6]
```diff
@@ -7,16 +7,18 @@ import (
...
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
 )
 
-// PoolAccountI defines an account interface for pools that hold tokens.
-type PoolAccountI interface {
-	authtypes.AccountI
+// PoolI defines an interface for pools that hold tokens.
+type PoolI interface {
+	proto.Message
```
Comment (ValarDragon) [inline x/gamm/types/pool.go:18]: Why does it implement the proto.Message interface?

Comment (sunnya97) [reply x/gamm/types/pool.go:18]: All the interfaces in the SDK are supposed to implement proto.Message

Comment (sunnya97) [reply x/gamm/types/pool.go:18]: Examples:

https://github.com/cosmos/cosmos-sdk/blob/9fd866e3820b3510010ae172b682d71594cd8c14/crypto/types/types.go#L10

https://github.com/cosmos/cosmos-sdk/blob/9fd866e3820b3510010ae172b682d71594cd8c14/crypto/types/types.go#L10

https://github.com/cosmos/cosmos-sdk/blob/9fd866e3820b3510010ae172b682d71594cd8c14/x/evidence/exported/evidence.go#L12

File: x/gamm/keeper/pool.go [hunk 7]
```diff
@@ -1,70 +1,103 @@
...
 }
 
-func (k Keeper) SetPool(ctx sdk.Context, poolAcc types.PoolAccountI) error {
+func (k Keeper) SetPool(ctx sdk.Context, pool types.PoolI) error {
 	// Make sure that pool exists
-	_, err := k.GetPool(ctx, poolAcc.GetId())
+	// _, err := k.GetPool(ctx, pool.GetId())
+	// if err != nil {
+	// 	return err
+	// }
```
Comment (ValarDragon) [inline x/gamm/keeper/pool.go:49]: Delete deadcode?

File: x/gamm/types/key.go [hunk 8]
```diff
@@ -17,16 +17,23 @@ const (
 )
 
 var (
-	PoolAddressPrefix = []byte("gamm_liquidity_pool")
-	GlobalPoolNumber  = []byte("gamm_global_pool_number")
-	// Used for querying to paginate the registered pool numbers.
-	PaginationPoolNumbers = []byte("gamm_pool_numbers_pagination")
+	// KeyLastLockID defines key to store lock ID used by last
```
Comment (ValarDragon) [inline x/gamm/types/key.go:20]: Incorrect comment

File: x/gamm/types/key.go [hunk 9]
```diff
@@ -17,16 +17,23 @@ const (
...
+	KeyGlobalPoolNumber = []byte{0x01}
+	// KeyPrefixPools defines prefix to store pools
+	KeyPrefixPools = []byte{0x02}
+
+	// // Used for querying to paginate the registered pool numbers.
+	// KeyPrefixPaginationPoolNumbers = []byte{0x03}
```
Comment (ValarDragon) [inline x/gamm/types/key.go:26]: Deadcode

File: x/gamm/types/key.go [hunk 10]
```diff
@@ -17,16 +17,23 @@ const (
...
 
-func GetKeyPaginationPoolNumbers(poolId uint64) []byte {
-	return append(PaginationPoolNumbers, sdk.Uint64ToBigEndian(poolId)...)
+func GetKeyPrefixPools(poolId uint64) []byte {
+	return append(KeyPrefixPools, sdk.Uint64ToBigEndian(poolId)...)
 }
+
+// func GetKeyPaginationPoolNumbers(poolId uint64) []byte {
+// 	return append(KeyPrefixPaginationPoolNumbers, sdk.Uint64ToBigEndian(poolId)...)
+// }
```
Comment (ValarDragon) [inline x/gamm/types/key.go:39]: Deadcode

**Review by ValarDragon — APPROVED** [review]
LGTM once remaining comments are resolved

File: x/gamm/types/pool.go (hunk 6 above)
Comment (ValarDragon) [reply x/gamm/types/pool.go:18]: ah I see. It literally just implements clone

**Pull request merged**